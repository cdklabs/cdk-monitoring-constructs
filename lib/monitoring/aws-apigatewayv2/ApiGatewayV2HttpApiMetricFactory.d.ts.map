{"version":3,"file":"ApiGatewayV2HttpApiMetricFactory.d.ts","sourceRoot":"","sources":["ApiGatewayV2HttpApiMetricFactory.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,QAAQ,EAAE,MAAM,iCAAiC,CAAC;AAC3D,OAAO,EAAE,aAAa,EAAE,MAAM,4BAA4B,CAAC;AAE3D,OAAO,EAGL,WAAW,EACX,aAAa,EAEb,qBAAqB,EACtB,MAAM,cAAc,CAAC;AAItB,MAAM,WAAW,qCAAqC;IACpD,QAAQ,CAAC,GAAG,EAAE,QAAQ,CAAC;IACvB;;OAEG;IACH,QAAQ,CAAC,QAAQ,CAAC,EAAE,MAAM,CAAC;IAC3B;;OAEG;IACH,QAAQ,CAAC,SAAS,CAAC,EAAE,MAAM,CAAC;IAC5B;;OAEG;IACH,QAAQ,CAAC,WAAW,CAAC,EAAE,MAAM,CAAC;IAC9B;;OAEG;IACH,QAAQ,CAAC,iBAAiB,CAAC,EAAE,OAAO,CAAC;IACrC;;OAEG;IACH,QAAQ,CAAC,qBAAqB,CAAC,EAAE,qBAAqB,CAAC;CACxD;AAED,qBAAa,gCAAgC;IAC3C,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;IAChD,SAAS,CAAC,QAAQ,CAAC,iBAAiB,EAAE,OAAO,CAAC;IAC9C,SAAS,CAAC,QAAQ,CAAC,qBAAqB,EAAE,qBAAqB,CAAC;IAChE,SAAS,CAAC,QAAQ,CAAC,aAAa,EAAE,aAAa,CAAC;gBAG9C,aAAa,EAAE,aAAa,EAC5B,KAAK,EAAE,qCAAqC;IAc9C;;OAEG;IACH,SAAS;IAST,oBAAoB;IASpB,qBAAqB;IAWrB,cAAc;IAWd,aAAa;IAUb,cAAc;IAWd,aAAa;IAUb;;OAEG;IACH,wBAAwB;IAIxB;;OAEG;IACH,wBAAwB;IAIxB;;OAEG;IACH,wBAAwB;IAIxB;;OAEG;IACH,mCAAmC;IAInC;;OAEG;IACH,mCAAmC;IAInC;;OAEG;IACH,mCAAmC;IAInC,gCAAgC,CAAC,WAAW,EAAE,WAAW;IAYzD,qBAAqB,CAAC,WAAW,EAAE,WAAW;CAW/C"}