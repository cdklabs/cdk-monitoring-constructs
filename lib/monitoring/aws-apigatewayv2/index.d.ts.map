{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,oCAAoC,CAAC;AACnD,cAAc,iCAAiC,CAAC"}