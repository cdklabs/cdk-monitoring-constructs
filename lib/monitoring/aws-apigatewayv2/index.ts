export * from "./ApiGatewayV2HttpApiMetricFactory";
export * from "./ApiGatewayV2HttpApiMonitoring";
