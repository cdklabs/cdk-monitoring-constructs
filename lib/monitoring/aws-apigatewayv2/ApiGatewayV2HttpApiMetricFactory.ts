import { IHttpApi } from "monocdk/aws-apigatewayv2";
import { DimensionHash } from "monocdk/aws-cloudwatch";

import {
  MetricFactory,
  MetricStatistic,
  RateComputationMethod,
} from "../../common";

const ApiGatewayNamespace = "AWS/ApiGateway";

export interface ApiGatewayV2HttpApiMetricFactoryProps {
  readonly api: IHttpApi;
  /**
   * @default $default
   */
  readonly apiStage?: string;
  /**
   * On undefined value is not set in dimensions
   */
  readonly apiMethod?: string;
  /**
   * On undefined value is not set in dimensions
   */
  readonly apiResource?: string;
  /**
   * @default true
   */
  readonly fillTpsWithZeroes?: boolean;
  /**
   * @default average
   */
  readonly rateComputationMethod?: RateComputationMethod;
}

export class ApiGatewayV2HttpApiMetricFactory {
  protected readonly metricFactory: MetricFactory;
  protected readonly fillTpsWithZeroes: boolean;
  protected readonly rateComputationMethod: RateComputationMethod;
  protected readonly dimensions: DimensionHash;

  constructor(
    metricFactory: MetricFactory,
    props: ApiGatewayV2HttpApiMetricFactoryProps
  ) {
    this.metricFactory = metricFactory;
    this.fillTpsWithZeroes = props.fillTpsWithZeroes ?? true;
    this.rateComputationMethod =
      props.rateComputationMethod ?? RateComputationMethod.AVERAGE;
    this.dimensions = {
      ApiId: props.api.httpApiId,
      Stage: props.apiStage ?? "$default",
      ...(props.apiMethod && { Method: props.apiMethod }),
      ...(props.apiResource && { Resource: props.apiResource }),
    };
  }

  metricTps() {
    // TODO: rename to metricInvocationRate and use rateComputationMethod
    return this.metricFactory.toRate(
      this.metricInvocationCount(),
      RateComputationMethod.PER_SECOND,
      false,
      "requests"
    );
  }

  metricInvocationCount() {
    return this.metricFactory.createMetric(
      "Count",
      MetricStatistic.SUM,
      "Invocations",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metric4xxCount() {
    return this.metricFactory.createMetric(
      "4xx",
      MetricStatistic.SUM,
      "4xx",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metric4xxRate() {
    const metric = this.metric4xxCount();
    return this.metricFactory.toRate(
      metric,
      this.rateComputationMethod,
      false,
      "errors"
    );
  }

  metric5xxCount() {
    return this.metricFactory.createMetric(
      "5xx",
      MetricStatistic.SUM,
      "5xx",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metric5xxRate() {
    const metric = this.metric5xxCount();
    return this.metricFactory.toRate(
      metric,
      this.rateComputationMethod,
      false,
      "faults"
    );
  }

  metricLatencyP50InMillis() {
    return this.metricFactory.createMetric(
      "Latency",
      MetricStatistic.P50,
      "P50 Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metricLatencyP90InMillis() {
    return this.metricFactory.createMetric(
      "Latency",
      MetricStatistic.P90,
      "P90 Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metricLatencyP99InMillis() {
    return this.metricFactory.createMetric(
      "Latency",
      MetricStatistic.P99,
      "P99 Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metricIntegrationLatencyP50InMillis() {
    return this.metricFactory.createMetric(
      "IntegrationLatency",
      MetricStatistic.P50,
      "P50 Integration Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metricIntegrationLatencyP90InMillis() {
    return this.metricFactory.createMetric(
      "IntegrationLatency",
      MetricStatistic.P90,
      "P90 Integration Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }

  metricIntegrationLatencyP99InMillis() {
    return this.metricFactory.createMetric(
      "IntegrationLatency",
      MetricStatistic.P99,
      "P99 Integration Latency",
      this.dimensions,
      undefined,
      ApiGatewayNamespace
    );
  }
}
