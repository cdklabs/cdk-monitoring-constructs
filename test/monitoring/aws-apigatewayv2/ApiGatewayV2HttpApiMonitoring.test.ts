import { Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { HttpApi } from "aws-cdk-lib/aws-apigatewayv2";

import {
  AlarmWithAnnotation,
  ApiGatewayV2HttpApiMonitoring,
} from "../../../lib";
import { addMonitoringDashboardsToStack } from "../../utils/SnapshotUtil";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  const testHttpApi = new HttpApi(stack, "testHttpApi", {
    apiName: "testHttpApi",
  });

  const monitoring = new ApiGatewayV2HttpApiMonitoring(scope, {
    api: testHttpApi,
    humanReadableName: "Dummy API Gateway for testing",
    alarmFriendlyName: "DummyApi",
  });

  addMonitoringDashboardsToStack(stack, monitoring);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  let numAlarmsCreated = 0;

  const testHttpApi = new HttpApi(stack, "testHttpApi", {
    apiName: "testHttpApi",
  });

  const monitoring = new ApiGatewayV2HttpApiMonitoring(scope, {
    api: testHttpApi,
    humanReadableName: "Dummy API Gateway for testing",
    alarmFriendlyName: "DummyApi",
    addLatencyP50Alarm: {
      Warning: {
        maxLatency: Duration.millis(150),
        datapointsToAlarm: 150,
      },
    },
    addLatencyP70Alarm: {
      Warning: {
        maxLatency: Duration.millis(170),
        datapointsToAlarm: 170,
      },
    },
    addLatencyP90Alarm: {
      Warning: {
        maxLatency: Duration.millis(190),
        datapointsToAlarm: 190,
      },
    },
    addLatencyP95Alarm: {
      Warning: {
        maxLatency: Duration.millis(195),
        datapointsToAlarm: 195,
      },
    },
    addLatencyP99Alarm: {
      Warning: {
        maxLatency: Duration.millis(199),
        datapointsToAlarm: 199,
      },
    },
    addLatencyP999Alarm: {
      Warning: {
        maxLatency: Duration.millis(1999),
        datapointsToAlarm: 1999,
      },
    },
    addLatencyP9999Alarm: {
      Warning: {
        maxLatency: Duration.millis(19999),
        datapointsToAlarm: 19999,
      },
    },
    addLatencyP100Alarm: {
      Warning: {
        maxLatency: Duration.millis(1100),
        datapointsToAlarm: 1100,
      },
    },
    addLatencyTM50Alarm: {
      Warning: {
        maxLatency: Duration.millis(250),
        datapointsToAlarm: 250,
      },
    },
    addLatencyTM70Alarm: {
      Warning: {
        maxLatency: Duration.millis(270),
        datapointsToAlarm: 270,
      },
    },
    addLatencyTM90Alarm: {
      Warning: {
        maxLatency: Duration.millis(290),
        datapointsToAlarm: 290,
      },
    },
    addLatencyTM95Alarm: {
      Warning: {
        maxLatency: Duration.millis(295),
        datapointsToAlarm: 295,
      },
    },
    addLatencyTM99Alarm: {
      Warning: {
        maxLatency: Duration.millis(299),
        datapointsToAlarm: 299,
      },
    },
    addLatencyTM999Alarm: {
      Warning: {
        maxLatency: Duration.millis(2999),
        datapointsToAlarm: 2999,
      },
    },
    addLatencyTM9999Alarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addLatencyTM95OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addLatencyTM99OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addLatencyTM999OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addLatencyTM9999OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addLatencyAverageAlarm: {
      Warning: {
        maxLatency: Duration.millis(20),
        datapointsToAlarm: 20,
      },
    },
    addIntegrationLatencyP50Alarm: {
      Warning: {
        maxLatency: Duration.millis(150),
        datapointsToAlarm: 150,
      },
    },
    addIntegrationLatencyP70Alarm: {
      Warning: {
        maxLatency: Duration.millis(170),
        datapointsToAlarm: 170,
      },
    },
    addIntegrationLatencyP90Alarm: {
      Warning: {
        maxLatency: Duration.millis(190),
        datapointsToAlarm: 190,
      },
    },
    addIntegrationLatencyP95Alarm: {
      Warning: {
        maxLatency: Duration.millis(195),
        datapointsToAlarm: 195,
      },
    },
    addIntegrationLatencyP99Alarm: {
      Warning: {
        maxLatency: Duration.millis(199),
        datapointsToAlarm: 199,
      },
    },
    addIntegrationLatencyP999Alarm: {
      Warning: {
        maxLatency: Duration.millis(1999),
        datapointsToAlarm: 1999,
      },
    },
    addIntegrationLatencyP9999Alarm: {
      Warning: {
        maxLatency: Duration.millis(19999),
        datapointsToAlarm: 19999,
      },
    },
    addIntegrationLatencyP100Alarm: {
      Warning: {
        maxLatency: Duration.millis(1100),
        datapointsToAlarm: 1100,
      },
    },
    addIntegrationLatencyTM50Alarm: {
      Warning: {
        maxLatency: Duration.millis(250),
        datapointsToAlarm: 250,
      },
    },
    addIntegrationLatencyTM70Alarm: {
      Warning: {
        maxLatency: Duration.millis(270),
        datapointsToAlarm: 270,
      },
    },
    addIntegrationLatencyTM90Alarm: {
      Warning: {
        maxLatency: Duration.millis(290),
        datapointsToAlarm: 290,
      },
    },
    addIntegrationLatencyTM95Alarm: {
      Warning: {
        maxLatency: Duration.millis(295),
        datapointsToAlarm: 295,
      },
    },
    addIntegrationLatencyTM99Alarm: {
      Warning: {
        maxLatency: Duration.millis(299),
        datapointsToAlarm: 299,
      },
    },
    addIntegrationLatencyTM999Alarm: {
      Warning: {
        maxLatency: Duration.millis(2999),
        datapointsToAlarm: 2999,
      },
    },
    addIntegrationLatencyTM9999Alarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addIntegrationLatencyTM95OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addIntegrationLatencyTM99OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addIntegrationLatencyTM999OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addIntegrationLatencyTM9999OutlierAlarm: {
      Warning: {
        maxLatency: Duration.millis(29999),
        datapointsToAlarm: 29999,
      },
    },
    addIntegrationLatencyAverageAlarm: {
      Warning: {
        maxLatency: Duration.millis(20),
        datapointsToAlarm: 20,
      },
    },
    add4xxCountAlarm: {
      Warning: {
        maxErrorCount: 0.02,
        datapointsToAlarm: 10,
      },
    },
    add4xxRateAlarm: {
      Warning: {
        maxErrorRate: 0.01,
        datapointsToAlarm: 20,
      },
    },
    add5xxCountAlarm: {
      Warning: {
        maxErrorCount: 2,
        datapointsToAlarm: 11,
      },
    },
    add5xxRateAlarm: {
      Warning: {
        maxErrorRate: 1,
        datapointsToAlarm: 22,
      },
    },
    addLowTpsAlarm: {
      Warning: { minTps: 1 },
    },
    addHighTpsAlarm: {
      Warning: { maxTps: 10 },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  addMonitoringDashboardsToStack(stack, monitoring);
  expect(numAlarmsCreated).toStrictEqual(46);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
