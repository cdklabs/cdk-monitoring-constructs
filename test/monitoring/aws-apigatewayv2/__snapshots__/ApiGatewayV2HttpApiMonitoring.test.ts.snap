// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Resources": Object {
    "ScopeTestDummyApiErrorCountWarning11BE36AF": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Error count is too high.",
        "AlarmName": "Test-DummyApi-Error-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 10,
        "EvaluationPeriods": 10,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "4xx",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "4xx",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 0.02,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiErrorRateWarningD181D1C6": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Error rate is too high.",
        "AlarmName": "Test-DummyApi-Error-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 20,
        "EvaluationPeriods": 20,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "4xx (avg)",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "4xx",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 0.01,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiFaultCountWarning6270BEC0": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Fault count is too high.",
        "AlarmName": "Test-DummyApi-Fault-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 11,
        "EvaluationPeriods": 11,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "5xx",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "5xx",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiFaultRateWarning63CFB7F0": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Fault rate is too high.",
        "AlarmName": "Test-DummyApi-Fault-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 22,
        "EvaluationPeriods": 22,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "5xx (avg)",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "5xx",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP50Critical4E10CCCB": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P50 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P50-Critical",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 11,
        "EvaluationPeriods": 11,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P50 Integration Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "IntegrationLatency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p50",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 110,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP50Warning7EEB3E57": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P50 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P50-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 11,
        "EvaluationPeriods": 11,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P50 Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "Latency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p50",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 110,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP90Critical72F6910F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P90 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P90-Critical",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 22,
        "EvaluationPeriods": 22,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P90 Integration Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "IntegrationLatency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p90",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 220,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP90Warning48639A66": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P90 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P90-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 22,
        "EvaluationPeriods": 22,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P90 Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "Latency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p90",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 220,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP99Critical6A9D549F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P99 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P99-Critical",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 33,
        "EvaluationPeriods": 33,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P99 Integration Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "IntegrationLatency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p99",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 330,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiLatencyP99Warning254DA12C": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "P99 latency is too high.",
        "AlarmName": "Test-DummyApi-Latency-P99-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 33,
        "EvaluationPeriods": 33,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "P99 Latency",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "Latency",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "p99",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 330,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiMaxTPSWarning62732F2B": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TPS is too high.",
        "AlarmName": "Test-DummyApi-MaxTPS-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "requests / PERIOD(requests)",
            "Id": "expr_1",
            "Label": "TPS",
          },
          Object {
            "Id": "requests",
            "Label": "Invocations",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "Count",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 10,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDummyApiMinTPSWarning4E8DB4D0": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TPS is too low.",
        "AlarmName": "Test-DummyApi-MinTPS-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "requests / PERIOD(requests)",
            "Id": "expr_1",
            "Label": "TPS",
          },
          Object {
            "Id": "requests",
            "Label": "Invocations",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "ApiId",
                    "Value": Object {
                      "Ref": "testHttpApiC57B300F",
                    },
                  },
                  Object {
                    "Name": "Stage",
                    "Value": "$default",
                  },
                ],
                "MetricName": "Count",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testHttpApiC57B300F": Object {
      "Properties": Object {
        "Name": "testHttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "testHttpApiDefaultStageBA851B68": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testHttpApiC57B300F",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Resources": Object {
    "testHttpApiC57B300F": Object {
      "Properties": Object {
        "Name": "testHttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "testHttpApiDefaultStageBA851B68": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testHttpApiC57B300F",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
}
`;
